### PR TITLE
Python: add parameter default value flow step

### PR DESCRIPTION
**Pull request by yoff**
This PR adds simple support for flow from default values to their function parameters.
The flow is added as a jump step rather than a local flow step, this has both reasons and consequences.
The reasons include that the default value is evaluated at initialisation time and so not in the same call context as the parameter. Making the flow step a jump step is an easy way to account for this. Consequences include that default values will not appear as local sources of the values observed inside the function body.

An alternative would be to create module variables (or something like those) to account for the phase transition between initialisation time and runtime. Then, at initialisation time, these would be written (presumably via jump steps, unless we have two flow analyses) and at runtime they would be read via local flow.

### CODE REVIEW

File: python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -1033,6 +1036,19 @@ predicate kwOverflowStoreStep(CfgNode nodeFrom, DictionaryElementContent c, Node
...
 }
 
+predicate defaultValueFlowStep(CfgNode nodeFrom, CfgNode nodeTo) {
+  exists(Function f, Parameter p, ParameterDefinition def |
+    // `getArgByName` supports, unlike `getAnArg`, keyword-only parameters
+    p = f.getArgByName(_) and
+    nodeFrom.asExpr() = p.getDefault() and
+    // The following expresses
+    // nodeTo.(ParameterNode).getParameter() = p
+    // without non-monotonic recursion
```
Comment (tausbn) [inline python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:1046]: ... So would that be "with monotonic recursion", then? :wink:

**Review by tausbn — CHANGES_REQUESTED** [review]
Looks good to me. It's _slightly_ awkward that accessing the default value won't be possible using `getALocalSource`, but I think it makes sense in this case (and the parameter _is_ available as a local source, so you can still get at the default value that way).

Just for completeness, I think this merits a performance comparison.

File: python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll (hunk 1 above)
Comment (yoff) [reply python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:1046]: ..the best kind of recursion :-)

**Comment by yoff** [comment]
> Just for completeness, I think this merits a performance comparison.

Agreed, experiment running [here](https://github.com/github/codeql-dca-main/tree/run/yoff/python-add-para__smoke-test__smoke-test/reports). (Be aware that the wizard told me: ```Found merge in the commit range. You are likely doing something wrong!
Please make sure to rebase your PR instead of using merge commits in the PR.``` Should we move to a rebase workflow? :p)

**Comment by yoff** [comment]
Evaluation seems ok..

**Comment by yoff** [comment]
> Evaluation seems ok..

Will evaluate using a larger set of queries and repos.

**Review by tausbn — APPROVED** [review]
Evaluation looks good. :shipit:

**Pull request merged**